### ISSUE TITLE
jwks json marshal

### ISSUE THREAD

**Issue by gbolo** [lestrrat-go/jwx#293]
I been using this library for testing an OIDC flow (as a client). I been using the jwa, jwt, and jwk packages you have published here to accomplish that. Works great, thank you.

I think it would be useful to be able to marshal the `jwk.Set` struct into an RFC compliant json structure, so that I can serve it directly as a jwks uri for signature verification. However, this struct is missing the json tag for the [`Keys` field](https://github.com/lestrrat-go/jwx/blob/v1.0.8/jwk/interface.go#L44), as a result the `json.Marshal` is producing non-compliant json ("Keys" instead of "keys"). https://tools.ietf.org/html/rfc7517#section-5.1

I am not sure if there already is an option to do this (I was unable to find one) in your extensive set of libraries. I think just adding the tag `json: "keys"` [here](https://github.com/lestrrat-go/jwx/blob/v1.0.8/jwk/interface.go#L44) is enough to satisfy this feature request.

Another handy feature would be to extend the `jwk.Set` with a method that would be able to produce a **PUBLIC KEY** version of your jwks in JSON format (in the case you have only private keys in it). For example:

```
# generate private key
rsaKey, _ := rsa.GenerateKey(rand.Reader, 2048)

# create a JWK from it
key, _ := jwk.New(rsaKey)
key.Set(jwk.KeyIDKey, "gbolo1")
key.Set(jwk.KeyUsageKey, jwk.ForSignature)
key.Set(jwk.AlgorithmKey, jwa.RS256)

# create a JWKS
jwkSet := &jwk.Set{Keys: []jwk.Key{key}}

# produce a jwks JSON with the public portions of my private key
jwksJsonBytes, err := jwkSet.PublicJSON()

...
# now i can serve this in my http handler so that my jwt tokens can get verified
http.HandleFunc("/jwks", func(w http.ResponseWriter, r *http.Request) {
   w.Write(jwksJsonBytes)
}

# this handler will produce json output like as follows:
## NOTICE: the private portion of the key(s) are omitted
{
  "keys": [
    {
      "use": "sig",
      "kty": "RSA",
      "kid": "gbolo1",
      "alg": "RS256",
      "n": "3YKkI...",
      "e": "AQAB"
    }
  ]
}


```

**Comment by lestrrat** [comment]
> ... is producing non-compliant json ("Keys" instead of "keys")

That's a bug :/
I've been meaning to do something about the clunkiness of `jwk.Set`, and maybe this is something that I should tackle in v1.1.0 (where I intend to introduce non-compatible changes). I just published an issue (#291) claiming "maybe April", though.

Is this a blocker for you, or are you able to workaround it for the time being? The changes itself is not that hard to do, I just want to coordinate the timing of breaking changes.

> Another handy feature would be to extend the jwk.Set with a method that would be able to produce a PUBLIC KEY version of your jwks in JSON format (in the case you have only private keys in it). 

That sounds like a useful tool, though if I incorporate it I would probably include it as a standalone function (i.e not a method)

```go
... = jwk.PublicKeySet(jwks)
```

**Comment by gbolo** [comment]
not a blocker at all. Current work around is something like:
```
type jwksJson struct {
	Keys []jwk.Key `json:"keys"`
}

func GetJwksNew() (encoded []byte) {
	encoded, _ = json.Marshal(jwksJson{Keys:pubKeySet.Keys})
	return
}
```

I really love the idea of the standalone function. Currently i maintain two different jwks objects:
```
var (
	signingKeySet *jwk.Set
	pubKeySet     *jwk.Set
)
```

Both vars have the same keys, except the `pubKeySet` is just the public portion of the keys in `signingKeySet`. This function you proposed would highly welcomed by me :)

If you would like me to contribute a PR I can look into it, just let me know (I will assume no for now). BTW, Thanks again for these great packages!

**Comment by lestrrat** [comment]
@gbolo I'm did some work the develop/v1.1.0 branch.

These are breaking changes, and API is not guaranteed to be stable yet, but I started work on the two things you metnioned in #296 and #297

For your `jwk.Set` marshaling, now it should Just Work. But the API is different, so you will see breakage.

For getting `jwk.Set` with public keys, this *should* do everything you want.

```go
set, err := jwk.Parse(jwksrc)
newSet, err := jwk.PublicSetOf(set)
```

Please review those changes, and let me know if they make sense. I'll merge them on to develop/v1.1.0 branch, and then if you wish, you can start using them (with caution) from that branch until we make a release

**Comment by gbolo** [comment]
@lestrrat I tested [topic/public-key-of](https://github.com/lestrrat-go/jwx/tree/topic/public-key-of) rev `3e59c0a` with the following code and it worked perfectly! Thank you:

```
pubKeySet, err = jwk.PublicSetOf(signingKeySet)
...
func GetJwksNew() (encoded []byte) {
	encoded, _ = json.Marshal(pubKeySet)
	return
}

# perfect json was produced:
{
  "keys": [
    {
      "alg": "RS256",
      "e": "AQAB",
      "kid": "1wIti9fhjOu7",
      "kty": "RSA",
      "n": "r0fm...",
      "use": "sig"
    },
    {
      "alg": "RS256",
      "e": "AQAB",
      "kid": "ckoeENvXW35v",
      "kty": "RSA",
      "n": "4cEH...",
      "use": "sig"
    }
  ]
}
```

I also noticed the changes to the API (now the `jwks.Set.Keys` is not exported). However, this was not a problem since the new API had everything I needed: `jwks.Set.Add(key)`, `jwks.Set.Len()`, and `jwks.Set.Get(int)`. I like the new API much better than before, excellent job.

**Comment by lestrrat** [comment]
Cool. Merged to develop/v1.1.0
